### PR TITLE
Creating APIs

### PR DESCRIPTION
**Pull request by Mohammad-Nayef**

### CODE REVIEW

File: RestaurantReservation.API/Extensions/ValidationResultExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿using FluentValidation.Results;
+
+namespace RestaurantReservation.API.Extensions
+{
+    public static class ValidationResultExtensions
+    {
+        /// <summary>
+        /// Returns only the important properties contained in the errors list.
+        /// </summary>
+        public static object GetImportantProperties(this ValidationResult validationResult)
```
Comment (OlaAtaiwi) [inline RestaurantReservation.API/Extensions/ValidationResultExtensions.cs]: define model to hold the result and return list of it (DONT define object as return type)

File: RestaurantReservation.API/Extensions/NamesValidationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        /// <summary>
+        /// Validate name property using FluentValidation.
+        /// </summary>
+        public static IRuleBuilderOptions<T, string> ValidateName<T>(
+            this IRuleBuilder<T, string> rule)
+        {
+            return rule
+                .Matches(@"^[A-Za-z\s]+$").WithMessage(
+                     "'{PropertyName}' should only contain letters and spaces.")
```
Comment (OlaAtaiwi) [inline RestaurantReservation.API/Extensions/NamesValidationExtensions.cs]: try to define all error messages as const var in separate file or in the same file (if u have few error codes)

File: RestaurantReservation.API/Extensions/ForeignKeyExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace RestaurantReservation.API.Extensions
+{
+    public static class ForeignKeyExtensions
+    {
+        /// <summary>
+        /// Extract the name of the wrong foreign key in the exception.
+        /// </summary>
+        public static string ExtractForeignKey(this DbUpdateException ex)
```
Comment (OlaAtaiwi) [inline RestaurantReservation.API/Extensions/ForeignKeyExtensions.cs]: hmmm what is the idea behind this?
we cannot guarantee that the exception always happens because of FK .

File: RestaurantReservation.API/Validators/CustomerWithoutIdValidator.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+                .ValidateName();
+
+            RuleFor(customer => customer.LastName)
+                .ValidateName();
+
+            RuleFor(customer => customer.Email)
+                .EmailAddress();
+
+            RuleFor(customer => customer.PhoneNumber)
+                .Matches(@"^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$");
```
Comment (OlaAtaiwi) [inline RestaurantReservation.API/Validators/CustomerWithoutIdValidator.cs]: this regex can be defined as const in the same file or another file containing other regexes.

Comment (OlaAtaiwi) [reply RestaurantReservation.API/Validators/CustomerWithoutIdValidator.cs]: and add msg for each validation and error codes

File: RestaurantReservation.API/Validators/UserLoginValidator.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace RestaurantReservation.API.Validators
+{
+    public class UserLoginValidator : AbstractValidator<UserLoginDTO>
+    {
+        public UserLoginValidator()
+        {
+            RuleFor(user => user.Username)
+                .Matches(@"[A-Za-z_]+").WithMessage("Only letters and underscores are allowed.")
+                .Length(1, 50);
```
Comment (OlaAtaiwi) [inline RestaurantReservation.API/Validators/UserLoginValidator.cs]:
```
RuleFor(user => user.Username)
                .Matches(@"[A-Za-z_]+")
                .WithMessage("Only letters and underscores are allowed.")
                .Length(1, 50);
```
define the regex in const or other file as well

File: RestaurantReservation/Program.cs [hunk 6]
```diff
@@ -5,128 +5,128 @@
 
 using var context = new RestaurantReservationDbContext();
```
Comment (OlaAtaiwi) [inline RestaurantReservation/Program.cs]: is this file needed?

File: RestaurantReservation.Db/Repositories/IEmployeeRepository.cs [hunk 7]
```diff
@@ -1,4 +1,4 @@
-﻿using RestaurantReservation.Db.Models;
+﻿using RestaurantReservation.Db.Entities;
 
 namespace RestaurantReservation.Db.Repositories
```
Comment (OlaAtaiwi) [inline RestaurantReservation.Db/Repositories/IEmployeeRepository.cs:3]: add summaries for all methods here and in other files

File: RestaurantReservation.API/Controllers/ReservationController.cs [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+            try
+            {
+                await _reservationService.UpdateAsync(reservationWithId);
+            }
+            catch (KeyNotFoundException)
+            {
+                return NotFound();
+            }
+            catch (DbUpdateException ex)
```
Comment (OlaAtaiwi) [inline RestaurantReservation.API/Controllers/ReservationController.cs]: we can not guarantee we would have only these exceptions
we have to add a catch for a general exception at the end

File: RestaurantReservation/Program.cs (hunk 6 above)
Comment (Mohammad-Nayef) [reply RestaurantReservation/Program.cs]: it was for the previous task, deleted it

File: RestaurantReservation.Db/Services/CustomerService.cs [hunk 9]
```diff
@@ -1,4 +1,4 @@
-﻿using RestaurantReservation.Db.Models;
+﻿using RestaurantReservation.Db.Entities;
 using RestaurantReservation.Db.Repositories;
 
 namespace RestaurantReservation.Db.Services
```
Comment (RandaZraik) [inline RestaurantReservation.Db/Services/CustomerService.cs:4]: All these services shouldn't be in the DB layer, we should introduce a new layer `RestaurantReservation.Application` that contains all these services and business logic, and this layer shouldn't have directly access to the db context, it will just be dealing with the repos through the interfaces, this way you are decoupling the business logic from the db layer.

File: RestaurantReservation.Db/Services/CustomerService.cs [hunk 10]
```diff
@@ -14,34 +14,34 @@ public CustomerService(RestaurantReservationDbContext context)
             _customersRepository = new(_context);
```
Comment (RandaZraik) [inline RestaurantReservation.Db/Services/CustomerService.cs:14]: Services should never have access to the db context at all, the idea of the repos interfaces is to use them to talk to DB without the need to directly know the implementation of DB layer nor the need to expose the db context to the other layers. 

Let's please remove this context from all services, and instead of `private CustomerRepository _customersRepository;` do `private ICustomerRepository _customersRepository;`. And you need to inject all repos with their interfaces into IoC as you did with the services. This way the repo constructor will automatically have the dbcontext without the need to pass down to it.

File: RestaurantReservation.API/Profiles/EmployeeProfile.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+using RestaurantReservation.API.Models;
+using RestaurantReservation.Db.Entities;
+
+namespace RestaurantReservation.API.Profiles
+{
+    public class EmployeeProfile : Profile
+    {
+        public EmployeeProfile()
+        {
+            CreateMap<Employee, EmployeeDTO>();
```
Comment (RandaZraik) [inline RestaurantReservation.API/Profiles/EmployeeProfile.cs:11]: We usually do not expose DB models to the API layer, if you want to map these models, profiles need to exist in the `Application` layer, and you can have these models in the `Application` layer instead of the API layer.

And to make that happen, we need to have API project to reference Application project, and DB project to reference Application project.

Lets actually need the new project Domain not Application. The idea here is that API and DB projects are frameworks/details that the domain project shouldn't rely on, it should be the way around, since domain layer usually refers to the business logic that should be pretty stable and won't require much changes in the future.

Having Domain to rely on DB layer for example is not a good idea, since DB is not stable as you could change the implementation of DB and use a different framework and Domain then will need to be updated as well. Domain should be stable and not affected as it's the business core of the project.

Read more about this, it's called Clean Architecture.

Comment (RandaZraik) [reply RestaurantReservation.API/Profiles/EmployeeProfile.cs:11]: Also, I see that DB models are returned from the repos, and the mapping happens in the API layer. As mentioned above, we don't expose DB models to the other layer and should be `internal` in the DB project. Instead, the repos should return domain/dto models,  and the mapping should happen from within the repos.

This mean the repo should only accept and return domain models/dtos, not db models.
